teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1071 @@
+    - name: myStep
+      image: myImage
+      #A step has a workspace information in
+      #its context so it can be used here directly
+      script: |
+        go test $(workspaces.shared.path)/pkg
+```
+</td>
+<td>
+
+```yaml
+# A standalone step has no context about a workspace.
+# It just needs a path to do its work 
+# which it can accept as a param.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: path
+      description: “Path to the test folder”
+  image: ...
+  script: |
+    go test $(params.path)
+
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  workspaces:
+    - name: shared
+  params:
+    - name: path
+      value: $(workspaces.shared.path)/pkg 
+      #A task has to provide a path to the 
+      # referenced step. Here, it’s done 
+      # via a workspace reference.
+  steps:
+    - name: myStep
+      stepRef:
+        name: myStep
+```
+</td>
+</tr>
+</table>
+
+While this a perfectly valid view point from the `Step` author’s perspective, the users using these `Steps` to construct `Tasks` and `Pipelines` will need to better understand how a path is used in the `Step` to provide an appropriate `Workspace`. We think that it’s better to start without `Workspaces` and see if this is causing major friction to users before introducing it in the `StepActions's` API spec.
+
+#### VolumeMounts
+In many cases, depending on the image used or the work that the `Step` is performing, a `VolumeMount` is extremely necessary. This is evident from the number of [catalog Tasks](https://github.com/search?q=repo%3Atektoncd%2Fcatalog%20volumeMounts&type=code) that use it. For standalone `StepActions`, it is challenging for authors to convey that such a `VolumeMount` is required for the `StepAction` to run successfully. Failure to provide this ability would make using the `StepAction` prone to errors and degrade user experience. 
+
+A `Task` could be using multiple `Steps` that require the same `VolumeMount`. When inlining the `Steps`, the `Task` author has the complete context and so this is easy for the author to avoid name clashes and confusion. However, `StepActions` could choose names that conflict with other `StepActions` because they are written in for general use cases. Because of this, we propose parameterizing the name of the `VolumeMount` when requested in a `StepAction`. This is something that can also be enforced during validation. 
+
+The `Task` authors can provide the underlying `Volume` (`emptyDir`, `Secret` etc.)
+
+<table>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: registryConfig
+  volumeMounts:
+    - name: $(params.registryConfig)
+       mountPath: /registry-config
+  image: ...
+  script: ...
+```
+</td>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: task-name
+spec:
+  volume:
```

Comment:
This was based on a lot of catalog Tasks that rely on volumeMounts. I think they are not deprecated but there was an effort trying to deprecate it. I don't think we got around to it. @dibyom had very good reasons for bringing this into `StepActions` so I want to let him explain why.

Reply:
This is a bit tricky - there are some volume/volumeMount use cases that are not covered by Workspaces: https://github.com/tektoncd/community/blob/main/teps/0112-replace-volumes-with-workspaces.md#volume-use-cases-supported-by-workspaces

We had considered adding functionality to workspaces for some of these use cases but decided not to do so given that the alternatives would be duplicating functionality or clunkier to use.

Without the `volumeMount` field, `StepAction` authors lose the ability to control where exactly a volume/workspace/disk/filesystem is mounted which may be necessary for the step to function correctly.  (The actual volume/disk that is mounted could come from a Pod volume or a workspace). 

For example, the cache needs to be mounted at `/workspace/.m2` for https://github.com/tektoncd/catalog/blob/537265eb583836f596f2a3a3ab3e3b904ae24013/task/jib-maven/0.4/jib-maven.yaml#L70-L73 and the certs need to be mounted at `/certs/client` for https://github.com/tektoncd/catalog/blob/537265eb583836f596f2a3a3ab3e3b904ae24013/task/docker-build/0.1/docker-build.yaml#L67-L69

Reply:
If a tool does not allow specifying the path to a folder they need it could be argued that's an issue in the tool. That said, I agree that, as Tekton, we may need to help users cope with those cases. 
I don't think we should promote a syntax where the `Task` author needs to use `volume` instead of `workspace` because that brings a runtime value (the actual name of the volume) into the authoring-time space.

If a `StepAction` needs a volume to be mounted at a special place, it could expose a parameter that provides the name of the volume, so that the step may additionally mount that volume on a different path. 

```yaml
apiVersion: tekton.dev/v1alpha1
kind: StepAction
metadata:
  name: myStep
spec:
  params:
    - name: registryConfigVolume
      description: the name of the volume with the registry configuration
  volumeMounts:
    - name: $(params.registryConfigVolume)
      mountPath: /registry-config
  image: ...
  script: ...
---
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: task-name
spec:
  workspaces:
  - name: registry-config
  steps:
  - name: myStep
    ref:
       name:  myStep
     params:
     - name: registryConfigVolume
       value: $(workspaces.registry-config.volume)
```

In short, I'm ok with the `volumeMounts` field in steps as long as we don't suggest it be used together with `volume` in `Task`. Another possibility, for a more concise and specific syntax, could be:

```yaml
apiVersion: tekton.dev/v1alpha1
kind: StepAction
metadata:
  name: myStep
spec:
  volumeMounts:
    - name: registryConfig
      description: the name of the volume with the registry configuration
      mountPath: /registry-config
  image: ...
  script: ...
---
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: task-name
spec:
  workspaces:
  - name: registry-config
  steps:
  - name: myStep
    ref:
       name:  myStep
     volumeMounts:
     - name: registryConfig
       volume: $(workspaces.registry-config.volume)
```

Reply:
Thanks @afrittoli I think in the example, I basically did a similar thing as your first approach. We parametrize the name of the volume. 

We can enforce that it is parametrized in a StepAction during validation. That way, we can make sure that the Task author can provide the appropriate tool (workspace or volume) during runtime.  I will capture this in the TEP.

Reply:
I updated the example to reflect your concerns and also noted that we **will** enforce that the name of the volumeMount is parametrized so that the Task can provide what it wants to during runtime.

Reply:
Thanks for updating the example